### PR TITLE
[SPARK-29977][SQL] Remove newMutableProjection/newOrdering/newNaturalAscendingOrdering  from SparkPlan

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This is to refactor `SparkPlan` code; it mainly removed `newMutableProjection`/`newOrdering`/`newNaturalAscendingOrdering` from `SparkPlan`.
The other modifications are listed below;
 - Move `BaseOrdering` from `o.a.s.sqlcatalyst.expressions.codegen.GenerateOrdering.scala` to `o.a.s.sqlcatalyst.expressions.ordering.scala`
 - `RowOrdering` extends `CodeGeneratorWithInterpretedFallback ` for `BaseOrdering`
 - Remove the unused variables (`subexpressionEliminationEnabled` and `codeGenFallBack`) from `SparkPlan`

### Why are the changes needed?

For better code/test coverage.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Existing.

### CODE REVIEW

**Comment by maropu** [comment]
@cloud-fan @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala [hunk 1]
```diff
@@ -102,4 +112,26 @@ object RowOrdering {
...
+
+  override protected def createCodeGeneratedObject(in: Seq[SortOrder]): BaseOrdering = {
+    GenerateOrdering.generate(in)
+  }
+
+  override protected def createInterpretedObject(in: Seq[SortOrder]): BaseOrdering = {
+    new InterpretedOrdering(in)
+  }
+
+  def create(order: Seq[SortOrder], inputSchema: Seq[Attribute]): Ordering[InternalRow] = {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: BaseOrdering?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala [hunk 2]
```diff
@@ -102,4 +112,26 @@ object RowOrdering {
...
+    createObject(bindReferences(order, inputSchema))
+  }
+
+  /**
+   * Creates a row ordering for the given schema, in natural ascending order.
+   */
+  def createNaturalAscendingOrdering(dataTypes: Seq[DataType]): Ordering[InternalRow] = {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: ditto?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: ah, yes.

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala [hunk 3]
```diff
@@ -19,18 +19,28 @@ package org.apache.spark.sql.catalyst.expressions
...
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateOrdering
 import org.apache.spark.sql.types._
 
 
+/**
+ * Inherits some default implementation for Java from `Ordering[Row]`
+ */
+class BaseOrdering extends Ordering[InternalRow] {
+  def compare(a: InternalRow, b: InternalRow): Int = {
+    throw new UnsupportedOperationException
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: do we still need to default implementation?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: Ah, that's right. I removed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala [hunk 4]
```diff
@@ -19,18 +19,24 @@ package org.apache.spark.sql.catalyst.expressions
...
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.BindReferences.bindReferences
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateOrdering
 import org.apache.spark.sql.types._
 
 
+/**
+ * A base class for generated/interpreted row ordering.
+ */
+abstract class BaseOrdering extends Ordering[InternalRow]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: hmm, so this is basically a type alias. Is it really useful?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala [hunk 5]
```diff
@@ -93,7 +93,7 @@ case class SortAggregateExec(
           initialInputBufferOffset,
           resultExpressions,
           (expressions, inputSchema) =>
-            newMutableProjection(expressions, inputSchema, subexpressionEliminationEnabled),
+            MutableProjection.create(expressions, inputSchema),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala]: Seems the new code path always set `subexpressionEliminationEnabled` to false. Can we provide a `create` method that takes `subexpressionEliminationEnabled` parameter?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala (hunk 4 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: Honestly, not sure. It seems this base class has been implemented for making it easily accessible from generated code? https://github.com/apache/spark/blob/d555f8fcc964e6e81f2d849de62d06877dca70c5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala#L158

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala (hunk 5 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala]: But, it seems `newMutableProjection` always ignores the parameter?
https://github.com/apache/spark/pull/26615/files#diff-b9f96d092fb3fea76bcf75e016799678L468

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala]: MutableProjection now directly gets the config from SQLConf.get.subexpressionEliminationEnabled.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala]: yea, I think so.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: ah i see, to work around type erasure

**Comment by maropu** [comment]
oh, I got the weird failures and that seems to come from the change https://github.com/apache/spark/pull/26615#discussion_r348907633.
I'll check now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala]: ah i see, thanks for confirmation!

**Comment by maropu** [comment]
It seems we need the default impl. for compare in `BaseOrdering`. That seems to be a janino restriction. I'm not exactly sure about why though, IIRC I saw the similar behaviour before....

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala (hunk 4 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: Ah, I see.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala [hunk 6]
```diff
@@ -19,18 +19,28 @@ package org.apache.spark.sql.catalyst.expressions
...
 import org.apache.spark.sql.catalyst.expressions.BindReferences.bindReferences
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateOrdering
 import org.apache.spark.sql.types._
 
 
+/**
+ * A base class for generated/interpreted row ordering.
+ */
+class BaseOrdering extends Ordering[InternalRow] {
+  def compare(a: InternalRow, b: InternalRow): Int = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: what if we only keep the method definition?
```
abstract class BaseOrdering extends Ordering[InternalRow] {
  def compare(a: InternalRow, b: InternalRow): Int
}
```

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: Let's check in the latest commit ;)
But, my guess is that we need to implement it based on the exception message:
```
File 'generated.java', Line 5, Column 1: Non-abstract class "org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificOrdering" must implement method "public abstract int scala.math.Ordering.compare(java.lang.Object, java.lang.Object)"
	at 
...
```

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ordering.scala]: Got the same exception;
```
Error running query: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 5, Column 1: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 5, Column 1: Non-abstract class "org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificOrdering" must implement method "public abstract int scala.math.Ordering.compare(java.lang.Object, java.lang.Object)"
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**